Realm/ObjectStore/impl/realm_coordinator.cpp

```diff
@@ -0,0 +1,427 @@
+
+RealmCoordinator::~RealmCoordinator()
+{
+    std::lock_guard<std::mutex> coordinator_lock(s_coordinator_mutex);
+    for (auto it = s_coordinators_per_path.begin(); it != s_coordinators_per_path.end(); ) {
+        if (it->second.expired()) {
+            it = s_coordinators_per_path.erase(it);
+        }
+        else {
+            ++it;
+        }
+    }
+}
+
+void RealmCoordinator::unregister_realm(Realm*)
+{
+    std::lock_guard<std::mutex> lock(m_realm_mutex);
+    for (size_t i = 0; i < m_cached_realms.size(); ++i) {
+        if (m_cached_realms[i].expired()) {
+            if (i + 1 < m_cached_realms.size()) {
+                m_cached_realms[i] = std::move(m_cached_realms.back());
+            }
+            m_cached_realms.pop_back();
+        }
+    }
+}
+
+void RealmCoordinator::clear_cache()
+{
+    std::vector<SharedRealm> realms_to_close;
+
+    {
+        std::lock_guard<std::mutex> lock(s_coordinator_mutex);
+
+        // Gather a list of all of the realms which will be removed
+        for (auto& weak_coordinator : s_coordinators_per_path) {
+            auto coordinator = weak_coordinator.second.lock();
+            if (!coordinator) {
+                continue;
+            }
+
+            for (auto& cached_realm : coordinator->m_cached_realms) {
+                if (auto realm = cached_realm.realm()) {
+                    realms_to_close.push_back(realm);
+                }
+            }
+        }
+
+        s_coordinators_per_path.clear();
+    }
+
+    // Close all of the previously cached Realms. This can't be done while
+    // s_coordinator_mutex is held as it may try to re-lock it.
+    for (auto& realm : realms_to_close) {
+        realm->close();
+    }
+}
+
+void RealmCoordinator::send_commit_notifications()
+{
+    m_notifier->notify_others();
+}
+
+void RealmCoordinator::pin_version(uint_fast64_t version, uint_fast32_t index)
+{
+    if (m_async_error) {
+        return;
+    }
+
+    SharedGroup::VersionID versionid(version, index);
+    if (!m_advancer_sg) {
+        try {
+            // Use a temporary Realm instance to open the shared group to reuse
+            // the error handling there
+            Realm tmp(m_config);
+            m_advancer_history = std::move(tmp.m_history);
+            m_advancer_sg = std::move(tmp.m_shared_group);
+            m_advancer_sg->begin_read(versionid);
+        }
+        catch (...) {
+            m_async_error = std::current_exception();
+            m_advancer_sg = nullptr;
+            m_advancer_history = nullptr;
+        }
+    }
+    else if (m_queries.empty() && m_new_queries.empty()) {
+        // If this is the first query then we don't already have a read transaction
+        m_advancer_sg->begin_read(versionid);
+    }
+    else if (versionid < m_advancer_sg->get_version_of_current_transaction()) {
+        // Ensure we're holding a readlock on the oldest version we have a
+        // handover object for, as handover objects don't
```

Comment:
I see now that this is likely a complete comment as there is an implied "hold a readlock".